src/mame/heathkit/tlb.cpp

```diff
@@ -0,0 +1,751 @@
+	// Original font dump
+	ROM_LOAD( "2716_444-29_h19font.bin", 0x0000, 0x0800, CRC(d595ac1d) SHA1(130fb4ea8754106340c318592eec2d8a0deaf3d0))
+
+	ROM_REGION( 0x1000, "keyboard", 0 )
+	// Original dump
+	ROM_LOAD( "2716_444-37_h19keyb.bin", 0x0000, 0x0800, CRC(5c3e6972) SHA1(df49ce64ae48652346a91648c58178a34fb37d3c))
+ROM_END
+
+ROM_START( watz19 )
+	// Watzman ROM
+	ROM_REGION( 0x2000, "maincpu", ROMREGION_ERASEFF )
+	ROM_LOAD( "watzman.bin", 0x0000, 0x1000, CRC(8168b6dc) SHA1(bfaebb9d766edbe545d24bc2b6630be4f3aa0ce9))
+
+	ROM_REGION( 0x0800, "chargen", 0 )
+	// Original font dump
+	ROM_LOAD( "2716_444-29_h19font.bin", 0x0000, 0x0800, CRC(d595ac1d) SHA1(130fb4ea8754106340c318592eec2d8a0deaf3d0))
+	// Watzman keyboard
+	ROM_REGION( 0x1000, "keyboard", 0 )
+	ROM_LOAD( "keybd.bin", 0x0000, 0x0800, CRC(58dc8217) SHA1(1b23705290bdf9fc6342065c6a528c04bff67b13))
+ROM_END
+
+ROM_START( ultra19 )
+	// ULTRA ROM
+	ROM_REGION( 0x2000, "maincpu", ROMREGION_ERASEFF )
+	ROM_LOAD( "2532_h19_ultra_firmware.bin", 0x0000, 0x1000, CRC(8ad4cdb4) SHA1(d6e1fc37a1f52abfce5e9adb1819e0030bed1df3))
+
+	ROM_REGION( 0x0800, "chargen", 0 )
+	// Original font dump
+	ROM_LOAD( "2716_444-29_h19font.bin", 0x0000, 0x0800, CRC(d595ac1d) SHA1(130fb4ea8754106340c318592eec2d8a0deaf3d0))
+	// Watzman keyboard
+	ROM_REGION( 0x1000, "keyboard", 0 )
+	ROM_LOAD( "2716_h19_ultra_keyboard.bin", 0x0000, 0x0800, CRC(76130c92) SHA1(ca39c602af48505139d2750a084b5f8f0e662ff7))
+ROM_END
+
+
+ioport_constructor heath_tlb_device::device_input_ports() const
+{
+	return INPUT_PORTS_NAME(tlb);
+}
+
+const tiny_rom_entry *heath_tlb_device::device_rom_region() const
+{
+	return ROM_NAME(h19);
+}
+
+void heath_tlb_device::serial_out_b(uint8_t data)
+{
+	m_write_sd(data);
+}
+
+WRITE_LINE_MEMBER(heath_tlb_device::cb1_w)
+{
+	m_ace->rx_w(state);
+}
+
+void heath_tlb_device::device_add_mconfig(machine_config &config)
+{
+	// basic machine hardware
+	Z80(config, m_maincpu, H19_CLOCK); // From schematics
+	m_maincpu->set_addrmap(AS_PROGRAM, &heath_tlb_device::mem_map);
+	m_maincpu->set_addrmap(AS_IO, &heath_tlb_device::io_map);
+
+	// video hardware
+	// TODO: make configurable, Heath offered 2 different CRTs - White, Green
+	screen_device &screen(SCREEN(config, "screen", SCREEN_TYPE_RASTER, rgb_t::green()));
+	screen.set_raw(MC6845_CLOCK * 45 / 2, 900, 260, 0, 640, 0, 250);
+	screen.set_screen_update("crtc", FUNC(mc6845_device::screen_update));
+
+	GFXDECODE(config, "gfxdecode", m_palette, gfx_h19);
+	PALETTE(config, "palette", palette_device::MONOCHROME);
+
+	MC6845(config, m_crtc, MC6845_CLOCK);
+	m_crtc->set_screen("screen");
+	m_crtc->set_show_border_area(true);
+	m_crtc->set_char_width(8);
+	m_crtc->set_update_row_callback(FUNC(heath_tlb_device::crtc_update_row));
+	m_crtc->out_vsync_callback().set_inputline(m_maincpu, INPUT_LINE_NMI); // frame pulse
+
+	// serial port
+	INS8250(config, m_ace, INS8250_CLOCK);
+	m_ace->out_int_callback().set_inputline("maincpu", INPUT_LINE_IRQ0);
```

Comment:
You can use object finders when hooking up callbacks rather than using literal tags – it’s less error-prone (i.e. you should use `m_maincpu` here rather than the literal tag).